"Send on closed channel" panic when trying to create a dedicated M2 host on aws

When creating a dedicated M2 host with the script (see bottom of he issue), one gets the following panic:

```json
$ ~/Downloads/mac/01_create-dedicatedhost.sh
690ba34fade920c9814a7768ace80dc0ad0a8c9ce275e46f15931e62cc925e9b
{"level":"debug","msg":"Running 'qenvs aws mac create'","time":"2023-12-06T11:32:27Z"}
{"level":"debug","msg":"context initialized for qenvs9ae01560","time":"2023-12-06T11:32:27Z"}
{"level":"debug","msg":"checking if m2 is offered at us-east-1","time":"2023-12-06T11:32:27Z"}
{"level":"debug","msg":"creating dedicated host for mac machine m2","time":"2023-12-06T11:32:29Z"}
{"level":"debug","msg":"Creating stack stackDedicatedHost-macmini","time":"2023-12-06T11:32:29Z"}
panic: send on closed channel

goroutine 85 [running]:
github.com/adrianriobo/qenvs/pkg/provider/aws/data.LokupRegionOfferingInstanceType.func1(0xc0004284e0)
	/workspace/pkg/provider/aws/data/instance-type.go:57 +0xd8
created by github.com/adrianriobo/qenvs/pkg/provider/aws/data.LokupRegionOfferingInstanceType
	/workspace/pkg/provider/aws/data/instance-type.go:53 +0x33e
```
The script 01_create-dedicatedhost.sh:
```bash
AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=
TARGET_REGION=us-east-1

podman run -d --name create-macmini --rm \
            -v ${PWD}:/workspace:z \
            -e AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID} \
            -e AWS_SECRET_ACCESS_KEY=${AWS_SECRET_ACCESS_KEY} \
            -e AWS_DEFAULT_REGION=${TARGET_REGION} \
            quay.io/rhqp/qenvs:v0.6.0 aws mac create \
                --project-name macmini \
                --backed-url s3://crcqe-tstenvs-tfstate/local/jsliacan/mac \
                --conn-details-output /workspace \
                --tags project=openshift-local,source=local,org=jsliacan \
                --only-host
              
# Check logs 
podman logs -f create-macmini
```

--- a/pkg/provider/aws/action/fedora/fedora.go
+++ b/pkg/provider/aws/action/fedora/fedora.go
@@ -62,7 +62,7 @@ func Create(r *Request) error {
 		r.spotPrice = so.MaxPrice
 	} else {
 		r.region = os.Getenv("AWS_DEFAULT_REGION")
-		az, err := data.GetRandomAvailabilityZone(r.region)
+		az, err := data.GetRandomAvailabilityZone(r.region, nil)
 		if err != nil {
 			return err
 		}

--- a/pkg/provider/aws/action/mac/mac-dh.go
+++ b/pkg/provider/aws/action/mac/mac-dh.go
@@ -6,7 +6,6 @@ import (
 	"github.com/adrianriobo/qenvs/pkg/manager"
 	qenvsContext "github.com/adrianriobo/qenvs/pkg/manager/context"
 	"github.com/adrianriobo/qenvs/pkg/provider/aws"
-	"github.com/adrianriobo/qenvs/pkg/provider/aws/data"
 	"github.com/adrianriobo/qenvs/pkg/provider/util/output"
 	"github.com/adrianriobo/qenvs/pkg/util/logging"
 	resourcesUtil "github.com/adrianriobo/qenvs/pkg/util/resources"
@@ -43,20 +42,20 @@ func (r *MacRequest) createDedicatedHost() (*string, *string, error) {
 // this function will create the dedicated host resource
 func (r *MacRequest) deployerDedicatedHost(ctx *pulumi.Context) (err error) {
 	ctx.Export(fmt.Sprintf("%s-%s", r.Prefix, outputRegion), pulumi.String(r.Region))
-	az, err := data.GetRandomAvailabilityZone(r.Region)
-	if err != nil {
-		return
-	}
+	// az, err := data.GetRandomAvailabilityZone(r.Region)
+	// if err != nil {
+	// 	return
+	// }
 	dh, err := ec2.NewDedicatedHost(ctx,
 		resourcesUtil.GetResourceName(r.Prefix, awsMacMachineID, "dh"),
 		&ec2.DedicatedHostArgs{
 			AutoPlacement:    pulumi.String("off"),
-			AvailabilityZone: pulumi.String(*az),
+			AvailabilityZone: pulumi.String(r.AvailabilityZone),
 			InstanceType:     pulumi.String(macTypesByArch[r.Architecture]),
 			Tags:             qenvsContext.ResourceTags(),
 		})
 	ctx.Export(fmt.Sprintf("%s-%s", r.Prefix, outputDedicatedHostID), dh.ID())
-	ctx.Export(fmt.Sprintf("%s-%s", r.Prefix, outputDedicatedHostAZ), pulumi.String(*az))
+	ctx.Export(fmt.Sprintf("%s-%s", r.Prefix, outputDedicatedHostAZ), pulumi.String(r.AvailabilityZone))
 	if err != nil {
 		return err
 	}

--- a/pkg/provider/aws/action/mac/mac.go
+++ b/pkg/provider/aws/action/mac/mac.go
@@ -43,6 +43,11 @@ func Create(r *MacRequest) (err error) {
 			return err
 		}
 		r.Region = *region
+		az, err := getAZ(r)
+		if err != nil {
+			return err
+		}
+		r.AvailabilityZone = *az
 		// No host id means need to create dedicated host
 		dhID, dhAZ, err := r.createDedicatedHost()
 		if err != nil {
@@ -100,7 +105,7 @@ func getRegion(r *MacRequest) (*string, error) {
 	logging.Debugf("checking if %s is offered at %s",
 		r.Architecture,
 		os.Getenv("AWS_DEFAULT_REGION"))
-	isOffered, err := data.IsInstaceTypeOffered(
+	isOffered, err := data.IsInstanceTypeOfferedByRegion(
 		macTypesByArch[r.Architecture],
 		os.Getenv("AWS_DEFAULT_REGION"))
 	if err != nil {
@@ -118,6 +123,26 @@ func getRegion(r *MacRequest) (*string, error) {
 	// We look for a region offering the type of instance
 	return data.LokupRegionOfferingInstanceType(
 		macTypesByArch[r.Architecture])
+}
+
+// Get a random AZ from the requested region, it ensures the az offers the instance type
+func getAZ(r *MacRequest) (az *string, err error) {
+	isOffered := false
+	var excludedAZs []string
+	for !isOffered {
+		az, err = data.GetRandomAvailabilityZone(r.Region, excludedAZs)
+		if err != nil {
+			return nil, err
+		}
+		isOffered, err = data.IsInstanceTypeOfferedByAZ(r.Region, macTypesByArch[r.Architecture], *az)
+		if err != nil {
+			return nil, err
+		}
+		if !isOffered {
+			excludedAZs = append(excludedAZs, *az)
+		}
+	}
+	return
 }
 
 // Mac machine can be dinamically moved across regions as it is

--- a/pkg/provider/aws/action/rhel/rhel.go
+++ b/pkg/provider/aws/action/rhel/rhel.go
@@ -82,7 +82,7 @@ func Create(r *Request) error {
 		r.spotPrice = so.MaxPrice
 	} else {
 		r.region = os.Getenv("AWS_DEFAULT_REGION")
-		az, err := data.GetRandomAvailabilityZone(r.region)
+		az, err := data.GetRandomAvailabilityZone(r.region, nil)
 		if err != nil {
 			return err
 		}

--- a/pkg/provider/aws/action/windows/windows.go
+++ b/pkg/provider/aws/action/windows/windows.go
@@ -94,7 +94,7 @@ func Create(r *Request) error {
 		r.spotPrice = so.MaxPrice
 	} else {
 		r.region = os.Getenv("AWS_DEFAULT_REGION")
-		az, err := data.GetRandomAvailabilityZone(r.region)
+		az, err := data.GetRandomAvailabilityZone(r.region, nil)
 		if err != nil {
 			return err
 		}

--- a/pkg/provider/aws/data/azs.go
+++ b/pkg/provider/aws/data/azs.go
@@ -11,11 +11,14 @@ import (
 	"golang.org/x/exp/slices"
 )
 
-func GetRandomAvailabilityZone(region string) (*string, error) {
+func GetRandomAvailabilityZone(region string, excludedAZs []string) (*string, error) {
 	azs, err := DescribeAvailabilityZones(region)
 	if err != nil {
 		return nil, err
 	}
+	azs = slices.DeleteFunc(azs, func(a *ec2.AvailabilityZone) bool {
+		return slices.Contains(excludedAZs, *a.ZoneName)
+	})
 	return azs[util.Random(len(azs)-1, 0)].ZoneName, nil
 }
 

--- a/pkg/provider/aws/data/instance-type.go
+++ b/pkg/provider/aws/data/instance-type.go
@@ -1,43 +1,73 @@
 package data
 
 import (
-	"github.com/adrianriobo/qenvs/pkg/util"
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/session"
-	"golang.org/x/exp/slices"
 
 	awsEC2 "github.com/aws/aws-sdk-go/service/ec2"
 )
 
+var (
+	locatioRegion string = "region"
+	locationAZ    string = "availability-zone"
+
+	filternameLocation     string = "location"
+	filternameInstanceType string = "instance-type"
+)
+
 // Get InstanceTypes offerings on current location
-func GetInstanceTypesOfferingsByRegion(region string) ([]string, error) {
+func IsInstanceTypeOfferedByRegion(instanceType, region string) (bool, error) {
 	config := aws.Config{}
 	if len(region) > 0 {
 		config.Region = aws.String(region)
 	}
 	sess, err := session.NewSession(&config)
 	if err != nil {
-		return nil, err
+		return false, err
 	}
 	svc := awsEC2.New(sess)
-	o, err := svc.DescribeInstanceTypeOfferings(nil)
-	if err != nil {
-		return nil, err
-	}
-	return util.ArrayConvert(o.InstanceTypeOfferings,
-			func(item *awsEC2.InstanceTypeOffering) string {
-				return *item.InstanceType
-			}),
-		nil
-}
-
-// Check if a instance type is available at the current location
-func IsInstaceTypeOffered(instanceType, region string) (bool, error) {
-	o, err := GetInstanceTypesOfferingsByRegion(region)
+	o, err := svc.DescribeInstanceTypeOfferings(
+		&awsEC2.DescribeInstanceTypeOfferingsInput{
+			LocationType: &locatioRegion,
+			Filters: []*awsEC2.Filter{
+				{
+					Name:   &filternameLocation,
+					Values: []*string{&region}},
+				{
+					Name:   &filternameInstanceType,
+					Values: []*string{&instanceType}},
+			}})
 	if err != nil {
 		return false, err
 	}
-	return slices.Contains(o, instanceType), nil
+	return len(o.InstanceTypeOfferings) == 1, nil
+}
+
+func IsInstanceTypeOfferedByAZ(region, instanceType, az string) (bool, error) {
+	config := aws.Config{}
+	if len(region) > 0 {
+		config.Region = aws.String(region)
+	}
+	sess, err := session.NewSession(&config)
+	if err != nil {
+		return false, err
+	}
+	svc := awsEC2.New(sess)
+	o, err := svc.DescribeInstanceTypeOfferings(
+		&awsEC2.DescribeInstanceTypeOfferingsInput{
+			LocationType: &locationAZ,
+			Filters: []*awsEC2.Filter{
+				{
+					Name:   &filternameLocation,
+					Values: []*string{&az}},
+				{
+					Name:   &filternameInstanceType,
+					Values: []*string{&instanceType}},
+			}})
+	if err != nil {
+		return false, err
+	}
+	return len(o.InstanceTypeOfferings) == 1, nil
 }
 
 // Check on all regions which offers the type of instance got one having it
@@ -51,7 +81,7 @@ func LokupRegionOfferingInstanceType(instanceType string) (*string, error) {
 	for _, region := range regions {
 		lRegion := region
 		go func(c chan string) {
-			if is, err := IsInstaceTypeOffered(
+			if is, err := IsInstanceTypeOfferedByRegion(
 				instanceType,
 				lRegion); err == nil && is {
 				c <- lRegion

--- a/pkg/provider/aws/data/instance-type.go
+++ b/pkg/provider/aws/data/instance-type.go
@@ -60,6 +60,5 @@ func LokupRegionOfferingInstanceType(instanceType string) (*string, error) {
 	}
 	// First region with offering is enoguh
 	oRegion := <-c
-	close(c)
 	return &oRegion, nil
 }

--- a/tkn/infra-aws-fedora.yaml
+++ b/tkn/infra-aws-fedora.yaml
@@ -4,7 +4,7 @@ kind: Task
 metadata:
   name: infra-aws-fedora
   labels:
-    app.kubernetes.io/version: "0.6.0"
+    app.kubernetes.io/version: "0.7.0-dev"
   annotations:
     tekton.dev/pipelines.minVersion: "0.24.x"
     tekton.dev/categories: infrastructure
@@ -77,7 +77,7 @@ spec:
 
   steps:
   - name: provisioner
-    image: quay.io/rhqp/qenvs:v0.6.0
+    image: quay.io/rhqp/qenvs:v0.7.0-dev
     imagePullPolicy: Always
     script: |
       #!/bin/sh

--- a/tkn/infra-aws-mac.yaml
+++ b/tkn/infra-aws-mac.yaml
@@ -4,7 +4,7 @@ kind: Task
 metadata:
   name: infra-aws-mac
   labels:
-    app.kubernetes.io/version: "0.6.0"
+    app.kubernetes.io/version: "0.7.0-dev"
   annotations:
     tekton.dev/pipelines.minVersion: "0.24.x"
     tekton.dev/categories: infrastructure
@@ -96,7 +96,7 @@ spec:
 
   steps:
   - name: provisioner
-    image: quay.io/rhqp/qenvs:v0.6.0
+    image: quay.io/rhqp/qenvs:v0.7.0-dev
     imagePullPolicy: Always
     script: |
       #!/bin/sh

--- a/tkn/infra-aws-rhel.yaml
+++ b/tkn/infra-aws-rhel.yaml
@@ -4,7 +4,7 @@ kind: Task
 metadata:
   name: infra-aws-rhel
   labels:
-    app.kubernetes.io/version: "0.6.0"
+    app.kubernetes.io/version: "0.7.0-dev"
   annotations:
     tekton.dev/pipelines.minVersion: "0.24.x"
     tekton.dev/categories: infrastructure
@@ -81,7 +81,7 @@ spec:
 
   steps:
   - name: provisioner
-    image: quay.io/rhqp/qenvs:v0.6.0
+    image: quay.io/rhqp/qenvs:v0.7.0-dev
     imagePullPolicy: Always
     script: |
       #!/bin/sh

--- a/tkn/infra-aws-windows-server.yaml
+++ b/tkn/infra-aws-windows-server.yaml
@@ -4,7 +4,7 @@ kind: Task
 metadata:
   name: infra-aws-windows-server
   labels:
-    app.kubernetes.io/version: "0.6.0"
+    app.kubernetes.io/version: "0.7.0-dev"
   annotations:
     tekton.dev/pipelines.minVersion: "0.24.x"
     tekton.dev/categories: infrastructure
@@ -86,7 +86,7 @@ spec:
 
   steps:
   - name: provisioner
-    image: quay.io/rhqp/qenvs:v0.6.0
+    image: quay.io/rhqp/qenvs:v0.7.0-dev
     imagePullPolicy: Always
     script: |
       #!/bin/sh

--- a/tkn/infra-azure-windows-desktop.yaml
+++ b/tkn/infra-azure-windows-desktop.yaml
@@ -4,7 +4,7 @@ kind: Task
 metadata:
   name: infra-azure-windows-desktop
   labels:
-    app.kubernetes.io/version: "0.6.0"
+    app.kubernetes.io/version: "0.7.0-dev"
   annotations:
     tekton.dev/pipelines.minVersion: "0.24.x"
     tekton.dev/categories: infrastructure
@@ -53,7 +53,7 @@ spec:
 
   steps:
   - name: provisioner
-    image: quay.io/rhqp/qenvs:v0.6.0
+    image: quay.io/rhqp/qenvs:v0.7.0-dev
     imagePullPolicy: Always
     volumeMounts:
     - mountPath: /opt/credentials/

